### PR TITLE
Add warnings when validation files are suspected to be identical with training files

### PR DESCRIPTION
**Pull request by haixiw**
*Description of changes:*
Add warnings when validation files are suspected to be identical with training files

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/sagemaker_xgboost_container/data_utils.py [hunk 1]
```diff
@@ -622,3 +622,39 @@ def _make_symlink(path, source_path, name, index):
...
+        if not os.path.exists(validate_path):
+            exc.UserError("validate path is not existed")
+
+    isRedundant = False
+    training_files_set = set(f for f in os.listdir(train_path) if os.path.isfile(os.path.join(train_path, f)))
+    validation_files_set = set(f for f in os.listdir(validate_path) if os.path.isfile(os.path.join(validate_path, f)))
+    same_name_files = training_files_set.intersection(validation_files_set)
+    if len(same_name_files) != 0:
+        for f in same_name_files:
+            f_train_name = os.path.join(train_path, f)
```
Comment (awsbmillare) [inline src/sagemaker_xgboost_container/data_utils.py:649]: nit: f_train_name is really f_train_path or like full_train_path, ditto for f_validate_name.

Also really f should be fname as its really a string (a name) and not a file object but I think that's fine here for concision.

File: src/sagemaker_xgboost_container/data_utils.py [hunk 2]
```diff
@@ -622,3 +622,39 @@ def _make_symlink(path, source_path, name, index):
...
+            f_validate_name = os.path.join(validate_path, f)
+            f_train_size = os.path.getsize(f_train_name)
+            f_validate_size = os.path.getsize(f_validate_name)
+            if f_train_size == f_validate_size:
+                isRedundant = True
+                logging.warning("Suspected identical files found. ({} and {} with same size {} bytes)."
+                                " Note: Duplicate data in the training set and validation set will"
```
Comment (awsbmillare) [inline src/sagemaker_xgboost_container/data_utils.py:656]: we can make a weaker claim here and just say "is usually not intentional and can impair validity of the model evaluation..."

File: src/sagemaker_xgboost_container/algorithm_mode/train.py [hunk 3]
```diff
@@ -93,6 +94,8 @@ def get_validated_dmatrices(train_path, validate_path, content_type, csv_weights
         if train_path == validate_path or os.path.basename(train_path) == os.path.basename(validate_path):
             logger.warning('Found same path for training and validation. This is not recommended and results may not '
                            'be correct.')
+        else:
```
Comment (awsbmillare) [inline src/sagemaker_xgboost_container/algorithm_mode/train.py:97]: for readability maybe put a comment here? that this just warns cx for now, no conditional logic happens here

**Review by awsbmillare — CHANGES_REQUESTED** [review]
Thanks for this work. It's great to help warn customers in this possibly common mistake case.

File: src/sagemaker_xgboost_container/algorithm_mode/train.py (hunk 3 above)
Comment (haixiw) [reply src/sagemaker_xgboost_container/algorithm_mode/train.py:97]: updated

File: test/unit/test_data_utils.py [hunk 4]
```diff
@@ -244,3 +244,8 @@ def test_parse_protobuf_dmatrix_single_feature_label(self):
                 pb_path = os.path.join(self.data_path, 'recordio_protobuf', file_path)
                 reader = data_utils.get_recordio_protobuf_dmatrix
                 self._check_dmatrix(reader, pb_path, 1, 1)
+
```
[the remark this answers is not in the document]
Comment (haixiw) [reply test/unit/test_data_utils.py:247]: added a negative case

**Review by awsbmillare — APPROVED** [review]

**Comment by awsbmillare** [comment]
why is there build failure? do we need to merge first to get to next stage so it builds correctly?

**Pull request merged**